compiler/rustc_middle/src/lint.rs

```diff
@@ -274,6 +274,44 @@ pub fn explain_lint_level_source(
     }
 }
 
+/// The innermost function for emitting lints.
+///
+/// If you are loocking to implement a lint, look for higher level functions,
+/// for example:
+/// - [`TyCtxt::emit_spanned_lint`]
+/// - [`TyCtxt::struct_span_lint_hir`]
+/// - [`TyCtxt::emit_lint`]
+/// - [`TyCtxt::struct_lint_node`]
+/// - `LintContext::lookup`
+///
+/// ## `decorate` signature
+///
+/// Signature of `decorate` may be confusing at first, for instance what's the
+/// point of returning `&'b mut DiagnosticBuilder<'a, ()>` if the original can
+/// be used instead?
+/// ```ignore pseudo-code
+/// _ = decorate(&mut diag);
+/// /* use `diag` here again */
+/// ```
```

Comment:
This is written from the perspective of whoever implements struct_lint_level, but the doc comment should be written for people that *call* `struct_lint_level`.
```suggestion
/// The return value of `decorate` is ignored by this function.So what is the
/// point of returning `&'b mut DiagnosticBuilder<'a, ()>`?
```